From 6572ab0d68c307354e619cc9e9d89531f1bd0c83 Mon Sep 17 00:00:00 2001
From: Huacai Chen <chenhc@lemote.com>
Date: Wed, 6 Nov 2013 15:19:01 +0800
Subject: [PATCH 034/130] MIPS: Loongson 3: Introduce disable_unused_cpus() to
 save power.

When use maxcpus=n to limit the number of cores, the initial offlined
cores are running a infinite loop in firmware and their clocks are not
disabled. For power saving, we take the unused cores up and then down
at boot time. When resume from STR, we do this the same.

Change-Id: I367f08942948c141211eeabb7687a92beb66879d
Signed-off-by: Huacai Chen <chenhc@lemote.com>
Signed-off-by: Hongliang Tao <taohl@lemote.com>
Signed-off-by: Hua Yan <yanh@lemote.com>
Signed-off-by: Zhang Shuangshuang <zhangshuangshuang@ict.ac.cn>
---
 arch/mips/include/asm/mach-loongson64/loongson.h |  6 ++++++
 arch/mips/loongson64/common/pm.c                 |  8 ++++++++
 arch/mips/loongson64/loongson-3/smp.c            | 18 ++++++++++++++++++
 3 files changed, 32 insertions(+)

diff --git a/arch/mips/include/asm/mach-loongson64/loongson.h b/arch/mips/include/asm/mach-loongson64/loongson.h
index aa165e51..19d713fd 100644
--- a/arch/mips/include/asm/mach-loongson64/loongson.h
+++ b/arch/mips/include/asm/mach-loongson64/loongson.h
@@ -360,4 +360,10 @@ extern unsigned long _loongson_addrwincfg_base;
 
 void loongson_suspend_lowlevel(void);
 
+#ifdef CONFIG_HOTPLUG_CPU
+extern int disable_unused_cpus(void);
+#else
+static inline int disable_unused_cpus(void) { return 0; }
+#endif
+
 #endif /* __ASM_MACH_LOONGSON64_LOONGSON_H */
diff --git a/arch/mips/loongson64/common/pm.c b/arch/mips/loongson64/common/pm.c
index f32df179..908d63f4 100644
--- a/arch/mips/loongson64/common/pm.c
+++ b/arch/mips/loongson64/common/pm.c
@@ -266,6 +266,13 @@ static int loongson_pm_begin(suspend_state_t state)
 	return 0;
 }
 
+static void loongson_pm_wake(void)
+{
+#ifdef CONFIG_CPU_LOONGSON3
+	disable_unused_cpus();
+#endif
+}
+
 static void loongson_pm_end(void)
 {
 #ifdef CONFIG_LOONGSON3_CPUAUTOPLUG
@@ -278,6 +285,7 @@ static const struct platform_suspend_ops loongson_pm_ops = {
 	.valid	= loongson_pm_valid_state,
 	.begin	= loongson_pm_begin,
 	.enter	= loongson_pm_enter,
+	.wake	= loongson_pm_wake,
 	.end	= loongson_pm_end,
 };
 
diff --git a/arch/mips/loongson64/loongson-3/smp.c b/arch/mips/loongson64/loongson-3/smp.c
index 7a335791..bb9dbda0 100644
--- a/arch/mips/loongson64/loongson-3/smp.c
+++ b/arch/mips/loongson64/loongson-3/smp.c
@@ -760,6 +760,24 @@ static int register_loongson3_notifier(void)
 }
 early_initcall(register_loongson3_notifier);
 
+int disable_unused_cpus(void)
+{
+	int cpu;
+	struct cpumask tmp;
+
+	cpumask_complement(&tmp, cpu_online_mask);
+	cpumask_and(&tmp, &tmp, cpu_possible_mask);
+
+	for_each_cpu(cpu, &tmp)
+		cpu_up(cpu);
+
+	for_each_cpu(cpu, &tmp)
+		cpu_down(cpu);
+
+	return 0;
+}
+core_initcall(disable_unused_cpus);
+
 #endif
 
 struct plat_smp_ops loongson3_smp_ops = {
-- 
2.11.0

